rskj-core/src/main/java/co/rsk/net/handler/quota/TxVirtualGasCalculator.java

```diff
@@ -0,0 +1,123 @@
+     *
+     * @param newTx      The tx being executed
+     * @param replacedTx The tx replaced by <code>newTx</code> if any
+     * @return The virtualGas consumed by the provided <code>newTx</code>
+     */
+    double calculate(Transaction newTx, @Nullable Transaction replacedTx) {
+        long txGasLimit = newTx.getGasLimitAsInteger().longValue();
+
+        long newTxNonce = newTx.getNonceAsInteger().longValue();
+        long futureNonceFactor = newTxNonce == accountNonce ? 1 : 2;
+
+        double lowGasPriceFactor = calculateLowGasPriceFactor(newTx);
+
+        double nonceFactor = 1 + NONCE_WEIGHT / (accountNonce + 1);
+
+        double sizeFactor = 1 + newTx.getSize() / SIZE_FACTOR_DIVISOR;
+
+        double replacementFactor = calculateReplacementFactor(newTx, replacedTx);
+
+        double gasLimitFactor = calculateGasLimitFactor(txGasLimit);
+
+        double compositeFactor = futureNonceFactor * lowGasPriceFactor * nonceFactor * sizeFactor * replacementFactor * gasLimitFactor;
+
+        double result = safeMultiply(txGasLimit, compositeFactor);
+
+        logger.trace("virtualGasConsumed calculation for tx [{}]: result = [{}] (txGasLimit {}, compositeFactor {}, futureNonceFactor {}, lowGasPriceFactor {}, nonceFactor {}, sizeFactor {}, replacementFactor {}, gasLimitFactor {})", newTx.getHash(), result, txGasLimit, compositeFactor, futureNonceFactor, lowGasPriceFactor, nonceFactor, sizeFactor, replacementFactor, gasLimitFactor);
+
+        return result;
+    }
+
+    private double calculateLowGasPriceFactor(Transaction newTx) {
+        long txGasPrice = newTx.getGasPrice().asBigInteger().longValue();
+
+        if (txGasPrice < this.avgGasPrice) {
+            double factor = (this.avgGasPrice - txGasPrice) / (this.avgGasPrice - (double) this.blockMinGasPrice);
+            return 1 + LOW_GAS_PRICE_WEIGH * factor;
+        } else {
+            return 1;
+        }
+    }
+
+    private double calculateReplacementFactor(Transaction newTx, @Nullable Transaction replacedTx) {
+        double newTxGasPrice = newTx.getGasPrice().asBigInteger().doubleValue();
+        double replacementRatio = Optional.ofNullable(replacedTx)
+                .map(Transaction::getGasPrice)
+                .map(rtxGasPrice -> rtxGasPrice.asBigInteger().doubleValue())
+                .map(rtxGasPrice -> newTxGasPrice / rtxGasPrice)
+                .orElse(0.0);
+        return replacementRatio > 0 ? (1 + 1 / replacementRatio) : 1;
+    }
+
+    private double calculateGasLimitFactor(long txGasLimit) {
+        return 1 + GAS_LIMIT_WEIGHT * txGasLimit / this.blockGasLimit;
+    }
+
+    @SuppressWarnings("squid:S1244")
+    private double safeMultiply(double op1, double op2) {
+        double result = op1 * op2;
+
+        if (Double.POSITIVE_INFINITY == result) {
+            throw new ArithmeticException("Overflow detected on virtualGas calculation");
```

Comment:
This is probably out of excess of caution as double is big enough to hold the multiplication being performed (txGasLimit is capped at 2**60 and the composite factor is capped at fairly less). 

In any case would change the exception for a return of, for example, the `txGasLimit * MAX_FACTOR` (that is, the maximum theoretical result), so capping the `result` of the operation between zero and txGasLimit * MAX_FACTOR

Reply:
I like your idea 👍

Reply:
Done.